distributed/scheduler.py

```diff
@@ -5139,59 +5172,74 @@ async def restart(self, client=None, timeout=30):
-                # naturally when they disconnect from the scheduler.
-            except TimeoutError:
-                logger.error(
-                    "Nannies didn't report back restarted within "
-                    "timeout.  Continuing with restart process"
-                )
-            else:
-                if not all(resp == "OK" for resp in resps):
-                    logger.error(
-                        "Not all workers responded positively: %s",
-                        resps,
-                        exc_info=True,
+            start = monotonic()
+            resps = await asyncio.gather(
+                *(
+                    asyncio.wait_for(
+                        # FIXME does not raise if the process fails to shut down,
+                        # see https://github.com/dask/distributed/pull/6427/files#r894917424
+                        # NOTE: Nanny will automatically restart worker process when it's killed
+                        nanny.kill(timeout=timeout),
+                        timeout,
                     )
+                    for nanny in nannies
+                ),
+                return_exceptions=True,
+            )
+            # NOTE: the `WorkerState` entries for these workers will be removed
+            # naturally when they disconnect from the scheduler.
 
-        self.clear_task_state()
+            # Remove any workers that failed to shut down, so we can guarantee
+            # that after `restart`, there are no old workers around.
+            bad_nannies = [
+                addr for addr, resp in zip(nanny_workers, resps) if resp is not None
+            ]
+            if bad_nannies:
+                await asyncio.gather(
+                    *(
+                        self.remove_worker(addr, stimulus_id=stimulus_id)
+                        for addr in bad_nannies
+                    )
+                )
 
-        self.erred_tasks.clear()
-        self.computations.clear()
+                raise TimeoutError(
+                    f"{len(bad_nannies)}/{len(nannies)} worker(s) did not shut down within {timeout}s"
+                )
 
         self.log_event([client, "all"], {"action": "restart", "client": client})
-        start = time()
-        while time() < start + 10 and len(self.workers) < n_workers:
-            await asyncio.sleep(0.01)
 
-        self.report({"op": "restart"})
+        if wait_for_workers:
+            while monotonic() < start + timeout:
+                if len(self.workers) >= n_workers:
+                    return
+                await asyncio.sleep(0.01)
+            else:
+                msg = (
+                    f"Waited for {n_workers} worker(s) to reconnect after restarting, "
+                    f"but after {timeout}s, only {len(self.workers)} have returned. "
+                    "Consider a longer timeout, or `wait_for_workers=False`."
+                )
+
+                if (n_nanny := len(nanny_workers)) < n_workers:
```

Comment:
I think we might also run into this condition if we have a mix of nannies and non-nannies (seems unlikely for any given deployment but possible) and the nannies are too slow on the restart.

Reply:
True. Not much we can do there though.

Reply:
I was thinking about adjusting the nanny to feature an event that we could await via RPC that would get set at the end of instantiation and unset when the worker process dies. That's for another PR though. This way, we could also refactor `Nanny.restart()` in general.